Internal: Filter sort by usage status used unused [ED-19153]

MuliElementor opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Add filter functionality to global classes list enabling sorting by usage status (used/unused classes)

Main changes:
- Implemented filter component with unused, empty, and on-page filter options
- Created SearchAndFilterProvider context to manage search and filter state
- Added filter logic to display classes based on usage status
- Refactored search component to use the new context-based approach

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

MuliElementor committed:
update

MuliElementor committed:
Add filtering functionality for CSS class usage in editor

This commit introduces hooks and components to enable filtering of CSS classes within the editor. It includes logic for filtering unused, empty, and page-relevant classes, alongside UI components like filters, chips, and clear actions for improved usability.

MuliElementor committed:
Refactor LabeledCheckbox and add NotFound config module.

Refactored `LabeledCheckbox` to improve code readability by adjusting spacing and formatting. Added a new `NotFound` configuration module to handle different "not found" states with a reusable structure, supporting scenarios like filtering, searching, or their combination.

MuliElementor committed:
Refactor and integrate search and filter contexts.

Replaced `CssClassNotFound` component with a more modular and reusable `NotFound` component. Introduced `SearchAndFilterProvider` to manage search and filter contexts across components, streamlining state management. Simplified and modularized related logic in class manager components for enhanced readability and maintainability.

MuliElementor committed:
Add filter and sort functionality to global classes management

This includes new context, types, and components for managing filters, such as "unused", "empty", and "on this page". Updated UI elements include a filter list with checkboxes and replaced the "Clear" icon with a "Brush" icon for reset functionality. Enhanced store selectors to support filtering unused and empty CSS classes.

MuliElementor committed:
Merge branch 'main' into ED-19153-filter-sort-by-usage-status-used-unused

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

MuliElementor committed:
udate

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

MuliElementor committed:
Merge remote-tracking branch 'origin/test-branch' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
Remove outdated context providers for filters and search

Replaced the `FilterAndSortContext` and `SearchContext` with a unified `SearchAndFilterProvider`, streamlining their usage. Improved code organization, removed abandoned components, and updated references to the new implementation while maintaining existing functionality.

MuliElementor committed:
Update packages/packages/core/editor-global-classes/src/components/css-class-usage/types.ts

Co-authored-by: gitstream-cm[bot] <111687743+gitstream-cm[bot]@users.noreply.github.com>

MuliElementor committed:
Update packages/packages/core/editor-global-classes/src/hooks/use-filters.ts

Co-authored-by: gitstream-cm[bot] <111687743+gitstream-cm[bot]@users.noreply.github.com>

MuliElementor committed:
Refactor code for improved readability and type usage

Reformatted code for consistent indentation and simplified type definitions. Replaced outdated import paths and streamlined inline type declarations for better maintainability. These changes enhance code clarity and adhere to best practices.

elementorbot committed:
Merge branch 'main' into test-branch

MuliElementor committed:
Update packages/packages/core/editor-global-classes/src/components/search-and-filter/context.tsx

Co-authored-by: gitstream-cm[bot] <111687743+gitstream-cm[bot]@users.noreply.github.com>

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

MuliElementor committed:
```
Add disableInteractive to Tooltips and fix async prefetch

Tooltips now include the disableInteractive prop to improve user experience by preventing unwanted interactions. Additionally, prefetchClassesUsage is updated to await its operation, ensuring proper asynchronous behavior.
```

MuliElementor committed:
Merge branch 'main' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
Merge remote-tracking branch 'origin/test-branch' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
lint

elementorbot committed:
Merge branch 'main' into test-branch

MuliElementor committed:
update test

MuliElementor committed:
Merge remote-tracking branch 'origin/test-branch' into ED-19153-filter-sort-by-usage-status-used-unused

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

elementorbot committed:
Merge branch 'main' into test-branch

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/filter-list.tsx:
```diff
@@ -0,0 +1,55 @@
+
+	return (
+		<MenuList>
+			{ checkBoxItems.map( ( { label, value }: CheckBoxItem ) => (
```
this iteration makes it hard to follow, maybe just render the 3 checkboxes it will be more clear

MuliElementor replied:
:\

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/labeld-checkbox.tsx:
```diff
@@ -0,0 +1,36 @@
+				checked={ checked }
+				sx={ {
+					padding: 0,
+					color: 'text.tertiary',
```
why do we need all those custom styles?

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/labeld-checkbox.tsx:
```diff
@@ -0,0 +1,36 @@
+	checked: boolean;
+};
+
+export const LabeledCheckbox = ( { label, suffix, onClick, checked }: LabeledCheckboxProps ) => (
```
since its only in use in 1 file, i think it should be there instead of a different file

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/context.tsx:
```diff
@@ -0,0 +1,76 @@
+		initialValue: '',
+	} );
+
+	const onClearSearch = useCallback( () => {
+		handleChange( '' );
+	}, [ handleChange ] );
```
```suggestion
	const onClearSearch = () => {
		handleChange( '' );
	};
```

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/context.tsx:
```diff
@@ -0,0 +1,76 @@
+		handleChange( '' );
+	}, [ handleChange ] );
+
+	const onClearFilter = React.useCallback( () => {
+		setFilters( INIT_CHECKED_FILTERS );
+	}, [] );
```
```suggestion
	const onClearFilter = () => {
		setFilters( INIT_CHECKED_FILTERS );
	};
```

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/context.tsx:
```diff
@@ -0,0 +1,76 @@
+					inputValue,
+					handleChange,
+					onClearSearch,
+					isSearchActive: inputValue.length >= 2,
```
i think this can be calculate in the component rather then the contect since its ui related and only i use there

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/clear-icon-button.tsx:
```diff
@@ -0,0 +1,20 @@
+
+type ClearIconButtonProps = { tooltipText: React.ReactNode; sxStyle?: SxProps< Theme > };
+
+export const ClearIconButton = ( { tooltipText, sxStyle }: ClearIconButtonProps ) => {
```
```suggestion
export const ClearIconButton = ( { tooltipText, sx }: ClearIconButtonProps ) => {
```

MuliElementor committed:
remove unused await

MuliElementor committed:
remove unused sx

MuliElementor committed:
remove useCallback and chane type

MuliElementor committed:
remove fragments

MuliElementor committed:
move function to one hook

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/components/css-class-usage-popover.tsx:
```diff
@@ -94,6 +94,7 @@ export const CssClassUsagePopover = ( {
 						<Stack flexDirection={ 'row' } flex={ 1 } alignItems={ 'center' }>
 							<Box display={ 'flex' } sx={ { pr: 1 } }>
 								<Tooltip
+									disableInteractive
```
removing interaction mean that the tooltip will not shown when the curser is not on the item

MuliElementor commented on packages/packages/core/editor-global-classes/src/hooks/use-filters.ts:
```diff
@@ -0,0 +1,29 @@
+	const allFilters = useFilteredCssClassUsage();
+
+	// Collect only the active filter keys
+	const activeKeys = useMemo(
+		() => Object.keys( filters ).filter( ( key ) => filters[ key as FilterKey ] ) as FilterKey[],
+		[ filters ]
+	);
+
+	return useMemo( () => {
+		if ( activeKeys.length === 0 ) {
+			return null;
+		}
+
+		// Start with the values of the first active filter
+		return activeKeys
+			.map( ( key ) => allFilters[ key ] || [] )
+			.reduce( ( acc, arr ) => acc.filter( ( val ) => arr.includes( val ) ) );
+	}, [ activeKeys, allFilters ] );
```
sure

MuliElementor committed:
Refactor filter components and optimize filter logic

Removed `LabeledCheckbox` from its own file and moved it inline where necessary to streamline code structure. Updated filter-related components to improve prop consistency (e.g., replacing `sxStyle` with `sx`). Enhanced filter logic by eliminating redundant operations and improving clarity in active filter handling.

MuliElementor committed:
Remove redundant `isSearchActive` property from context.

The `isSearchActive` property was replaced with a derived condition based on `inputValue.length >= 2`. This simplifies the context by avoiding duplication and ensures the search state is dynamically inferred directly from `inputValue`.

MuliElementor committed:
Refactor NotFound components into a centralized implementation

Consolidated the NotFound logic into a single, reusable component under 'not-found'. Removed redundant code and simplified its usage across components, improving maintainability and consistency.

MuliElementor committed:
Merge branch 'main' into test-branch

MuliElementor committed:
Refactor filter type definitions and usage

Moved `FilterKey` and `CheckedFilters` types to relevant hooks for better modularity. Updated filter handling logic to align with the new type locations, improving code clarity and maintainability.

MuliElementor committed:
Merge remote-tracking branch 'origin/test-branch' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
Merge branch 'main' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
Refactor checkbox items directly into the JSX structure

Removed the `CheckBoxItem` array and inlined its data directly into the JSX `LabeledCheckbox` components. This simplifies the component's structure and reduces unnecessary indirection.

MuliElementor committed:
reverty changes since its used other places

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/not-found/not-found.tsx:
```diff
@@ -0,0 +1,133 @@
+import * as React from 'react';
```
we can remove the folder as well

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/active-filters.tsx:
```diff
@@ -0,0 +1,46 @@
+		filters: { filters, setFilters },
+	} = useSearchAndFilters();
+
+	const handleRemoveFilter = React.useCallback(
```
i don't see any value for `useCallback` here, especially when you pass an arrow function later
as a rule of thumb we dont use memozations to much unless it needed

IshayMaya commented on packages/packages/core/editor-global-classes/src/components/search-and-filter/components/filter/active-filters.tsx:
```diff
@@ -0,0 +1,46 @@
+							<Chip
+								size="tiny"
+								key={ key }
+								label={ checkBoxItems.find( ( item ) => item.value === key )?.label }
```
if its only for labels think it'll be more efficient to just hold an object with key value pair

MuliElementor committed:
update

MuliElementor committed:
Refactor global classes manager and improve filter logic

Streamlined the Class Manager UI components, replaced redundant code with reusable structures, and introduced a total CSS class counter. Simplified filter management by using a centralized configuration and optimized active filter handling for better readability and maintainability.

MuliElementor committed:
remove folder

MuliElementor committed:
Merge branch 'main' into ED-19153-filter-sort-by-usage-status-used-unused

MuliElementor committed:
Update Chip and FilterIcon components to use 'tiny' size

Replaced 'small' size with 'tiny' for Chip components and added 'tiny' font size to FilterIcon. This ensures consistency in visual design and aligns with updated style guidelines.

MuliElementor committed:
Refactor TotalCssClassCounter for inline definition

Moved the TotalCssClassCounter component from its own file to an inline definition within class-manager-panel.tsx. This change simplifies the file structure and reduces unnecessary imports.

MuliElementor committed:
Refactor TotalCssClassCounter for inline definition

Moved the TotalCssClassCounter component from its own file to an inline definition within class-manager-panel.tsx. This change simplifies the file structure and reduces unnecessary imports.

MuliElementor committed:
Merge branch 'main' into test-branch

IshayMaya committed:
Merge branch 'main' into test-branch
